### PR TITLE
Support arbitrary length identifiers in Mach-O

### LINKED ISSUE #17056: Long identifiers in Mach-O are mishandled
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | MacOS 64-bit
| File format of the file you reverse (mandatory)      | Mach-O dynamic library
| Architecture/bits of the file (mandatory)            | arm64, arm64e
| r2 -v full output, **not truncated** (mandatory)     | radare2 4.5.0-git 24799 @ darwin-x86-64 git.4.4.0-274-g77cc18ba0 commit: 77cc18ba046bc1966704df759f37833a063f9e01 build: 2020-06-12__12:13:52

While working with some dylibs I noticed that long identifiers are being truncated. I created a tiny Mach-O dynamic library to investigate this behavior. The dylib exports 2 symbols, one is 1024 chars long and the other is 2048 chars long.

### Expected behavior

radare2 should be able to handle long identifier, similarly to how nm, `Apple LLVM version 11.0.3 (clang-1103.0.32.62)` handles them

```
$ nm longIdentifiers.dylib
0000000000007fa8 T _abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghij
0000000000007fb0 T _abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrst
                 U dyld_stub_binder
```

### Actual behavior

radare2 truncates one of the identifiers and ignores the other one

```
$ rabin2 -s longIdentifiers.dylib
[Symbols]

nth paddr       vaddr      bind   type size lib name
――――――――――――――――――――――――――――――――――――――――――――――――――――
0    0x00007fa8 0x00007fa8 GLOBAL FUNC 0        _abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrs
1    0x00007fa8 0x00007fa8 LOCAL  FUNC 0        func.00007fa8
2    0x00007fb0 0x00007fb0 LOCAL  FUNC 0        func.00007fb0
```

### Steps to reproduce the behavior 

```
rabin2 -s longIdentifiers.dylib
```
[longIdentifiers.zip](https://github.com/radareorg/radare2/files/4770242/longIdentifiers.zip)

### Additional Logs, screenshots, source-code,  configuration dump, ...

The following code in `mach0.c` is pretty staright forward.

```
2900	stridx = bin->symtab[i].n_strx;
2901	char *sym_name = get_name (bin, stridx, false);
2902	if (sym_name) {
2903		r_str_ncpy (symbols[j].name, sym_name, R_BIN_MACH0_STRING_LENGTH);
2904		free (sym_name);
2905	}
```

`get_name()` returns an `strdup`ed copy of the string, then the string is copied with max length of `R_BIN_MACH0_STRING_LENGTH`, into the relevant `symbol_t` and the `strdup`ed pointer is freed. Peeking in `mach0.h:11` we can see `#define R_BIN_MACH0_STRING_LENGTH 256`. 

Then, the code continues to

```
2920	if (inSymtab (hash, symbols[j].name, symbols[j].addr)) {
2921		symbols[j].name[0] = 0;
2922		j--;
2923	}
```

and `inSymtab()` returns `true` for the second identifier. Therefore, the second identifier is dropped form `bin->symbols` and not shown in `rabin2 -s`.

### Discussion

Initially I tried to solve it by changing `256` to `512` and for the purpose of my work, it did the job. But later, while thinking about how to fix this bug I searched for long identifiers in the Dyld Shared Cache and found out some extremely long identifiers, such as `__ZN3WTF18__destroy_op_tableINS_7VariantIJN7WebCore15CanvasDirectionENS2_14CanvasFillRuleENS2_13CanvasLineCapENS2_14CanvasLineJoinENS2_15CanvasTextAlignENS2_18CanvasTextBaselineENS2_21ImageSmoothingQualityENS2_15DOMMatrix2DInitEPNS2_7ElementEPNS2_16HTMLImageElementEPNS2_11ImageBitmapEPNS2_9ImageDataEPNS2_6Path2DEPNS2_11WebGLBufferEPNS2_16WebGLFramebufferEPNS2_12WebGLProgramEPNS2_10WebGLQueryEPNS2_17WebGLRenderbufferEPNS2_12WebGLSamplerEPNS2_11WebGLShaderEPNS2_9WebGLSyncEPNS2_12WebGLTextureEPNS2_22WebGLTransformFeedbackEPNS2_20WebGLUniformLocationEPNS2_22WebGLVertexArrayObjectENS_6RefPtrIN3JSC11ArrayBufferENS_13DumbPtrTraitsIS1B_EEEENS19_INS1A_15ArrayBufferViewENS1C_IS1F_EEEENS19_INS2_14CanvasGradientENS1C_IS1I_EEEENS19_INS2_13CanvasPatternENS1C_IS1L_EEEENS19_INS1A_21GenericTypedArrayViewINS1A_14Float32AdaptorEEENS1C_IS1Q_EEEENS19_INS2_17HTMLCanvasElementENS1C_IS1T_EEEENS19_ISD_NS1C_ISD_EEEENS19_INS2_16HTMLVideoElementENS1C_IS1Y_EEEENS19_ISF_NS1C_ISF_EEEENS19_INS2_20TypedOMCSSImageValueENS1C_IS23_EEEENS19_ISH_NS1C_ISH_EEEENS19_INS1O_INS1A_12Int32AdaptorEEENS1C_IS29_EEEENS19_INS1O_INS1A_13Uint32AdaptorEEENS1C_IS2D_EEEENS1_IJS1X_S1V_S22_S25_S20_EEENS1_IJNS_6StringES1K_S1N_EEENS1_IJS1E_S1H_EEENS_8OptionalIS2J_EENS1_IJS22_S27_S1X_S1V_S20_EEENS2K_IS2M_EENS_6VectorIS2H_Lm0ENS_15CrashOnOverflowELm16ENS_10FastMallocEEENS2O_IfLm0ES2P_Lm16ES2Q_EENS2O_IjLm0ES2P_Lm16ES2Q_EENS2O_IiLm0ES2P_Lm16ES2Q_EENS1_IJS1S_S2S_EEENS1_IJS2B_S2U_EEENS1_IJS2F_S2T_EEES2H_dfNS2K_IfEEyxjihbEEENS_16__index_sequenceIJLl0ELl1ELl2ELl3ELl4ELl5ELl6ELl7ELl8ELl9ELl10ELl11ELl12ELl13ELl14ELl15ELl16ELl17ELl18ELl19ELl20ELl21ELl22ELl23ELl24ELl25ELl26ELl27ELl28ELl29ELl30ELl31ELl32ELl33ELl34ELl35ELl36ELl37ELl38ELl39ELl40ELl41ELl42ELl43ELl44ELl45ELl46ELl47ELl48ELl49ELl50ELl51ELl52ELl53ELl54ELl55ELl56ELl57ELl58ELl59ELl60EEEEE14__destroy_funcILl48EEEvPS2Z_` of `WebCore`, which is 1836 chars long. Then, I went to look up if there's any hard upper boundry for identifiers lengths.

Given the fact that the C++ standard defines identifiers as an ["arbitrarily long sequence of letters and digits"](https://timsong-cpp.github.io/cppwp/lex.name#1), we should rethink about how we handle identifiers memory.

The only solution I can think of, that can handle arbitrary long identifiers, is changing `name` of `symbol_t` from `char[R_BIN_MACH0_STRING_LENGTH]` to `char *` and storing the pointer returned by `get_name()` there. It would require additional cleanup when `bin` object is destroyed, to avoid memory leaks.

### PR DESCRIPTION
**Pull request by phakeobj**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

I changed `name` member of `symbol_t` from `char[R_BIN_MACH0_STRING_LENGTH]` (256) to `char *` to support arbitrary length symbols. Since symbols are now allocated on the heap, I added cleanup in `destroy`.

**Test plan**

I created a Mach-O dylib with two identifiers, one 1024 chars long and the other 2048 chars long, and added a test case that expects `is` to output the two identifiers correctly (as compared with the output of `nm` which processes long identifiers successfully).

**Closing issues**

Closes #17056

### CODE REVIEW

File: libr/bin/format/mach0/mach0.c [hunk 1]
```diff
@@ -2050,6 +2050,12 @@ void *MACH0_(mach0_free)(struct MACH0_(obj_t) *mo) {
 	if (!mo) {
 		return NULL;
 	}
+
+	for (size_t i = 0; !mo->symbols[i].last; i++) {
```
Comment (trufae) [inline libr/bin/format/mach0/mach0.c:2054]: do not define variables inside the for loop block, this is part of the c99 standard we don't follow

File: libr/bin/format/mach0/mach0.c [hunk 2]
```diff
@@ -2050,6 +2050,12 @@ void *MACH0_(mach0_free)(struct MACH0_(obj_t) *mo) {
...
+		if (mo->symbols[i].name) {
```
Comment (trufae) [inline libr/bin/format/mach0/mach0.c:2055]: unnecessary conditional. free(NULL) is totally valid

File: libr/bin/format/mach0/mach0.c [hunk 3]
```diff
@@ -2414,7 +2420,7 @@ static bool parse_import_stub(struct MACH0_(obj_t) *bin, struct symbol_t *symbol
 
 static int inSymtab(HtPP *hash, const char *name, ut64 addr) {
 	bool found;
-	const char *key = sdb_fmt ("%s.%"PFMT64x, name, addr);
+	const char *key = sdb_fmt ("%"PFMT64x".%s", addr, name);
```
Comment (trufae) [inline libr/bin/format/mach0/mach0.c:2423]: sdb_fmt have a very low limit . better use r_str_newf

File: libr/bin/format/mach0/mach0.c [hunk 4]
```diff
@@ -2898,15 +2904,12 @@ const struct symbol_t *MACH0_(get_symbols)(struct MACH0_(obj_t) *bin) {
...
 				stridx = bin->symtab[i].n_strx;
-				char *sym_name = get_name (bin, stridx, false);
-				if (sym_name) {
-					r_str_ncpy (symbols[j].name, sym_name, R_BIN_MACH0_STRING_LENGTH);
-					free (sym_name);
-				}
-				symbols[j].name[R_BIN_MACH0_STRING_LENGTH - 2] = 0;
+				symbols[j].name = get_name (bin, stridx, false);
+				// symbols[j].name[R_BIN_MACH0_STRING_LENGTH - 2] = 0;
 				symbols[j].last = 0;
```
Comment (trufae) [inline libr/bin/format/mach0/mach0.c:2909]: last should be bool. can you do the change in this PR too? thanks!

File: libr/bin/format/mach0/mach0.c [hunk 5]
```diff
@@ -2954,26 +2958,26 @@ const struct symbol_t *MACH0_(get_symbols)(struct MACH0_(obj_t) *bin) {
 				}
 				char *sym_name = get_name (bin, stridx, false);
 				if (sym_name) {
-					r_str_ncpy (symbols[j].name, sym_name, R_BIN_MACH0_STRING_LENGTH);
-					free (sym_name);
+					symbols[j].name = sym_name;
 				} else {
-					sprintf (symbols[j].name, "entry%d\n", i);
+					symbols[j].name = r_str_new ("entry%d\n");
```
Comment (trufae) [inline libr/bin/format/mach0/mach0.c:2963]: missing argument to format string. use r_str_newf instead or strdup

File: libr/bin/format/mach0/mach0.c [hunk 6]
```diff
@@ -2954,26 +2958,26 @@ const struct symbol_t *MACH0_(get_symbols)(struct MACH0_(obj_t) *bin) {
...
 					//symbols[j].name[0] = 0;
 				}
-				symbols[j].name[R_BIN_MACH0_STRING_LENGTH - 1] = 0;
 				symbols[j].last = 0;
 				if (inSymtab (hash, symbols[j].name, symbols[j].addr)) {
-					symbols[j].name[0] = 0;
+					free (symbols[j].name);
+					symbols[j].name = 0;
```
Comment (trufae) [inline libr/bin/format/mach0/mach0.c:2969]: = NULL

File: libr/bin/format/mach0/mach0.c [hunk 7]
```diff
@@ -2898,15 +2902,12 @@ const struct symbol_t *MACH0_(get_symbols)(struct MACH0_(obj_t) *bin) {
...
-				if (sym_name) {
-					r_str_ncpy (symbols[j].name, sym_name, R_BIN_MACH0_STRING_LENGTH);
-					free (sym_name);
-				}
-				symbols[j].name[R_BIN_MACH0_STRING_LENGTH - 2] = 0;
-				symbols[j].last = 0;
-				if (bin->main_addr == 0) {
-					const char *name = symbols[j].name;
+				symbols[j].name = get_name (bin, stridx, false);
+				// symbols[j].name[R_BIN_MACH0_STRING_LENGTH - 2] = 0;
```
Comment (ret2libc) [inline libr/bin/format/mach0/mach0.c:2906]: remove commented code.

File: libr/bin/p/bin_mach0.c [hunk 8]
```diff
@@ -213,7 +213,7 @@ static RList *symbols(RBinFile *bf) {
 	Sdb *symcache = sdb_new0 ();
 	bin = (struct MACH0_(obj_t) *) obj->bin_obj;
 	for (i = 0; !syms[i].last; i++) {
-		if (!syms[i].name[0] || syms[i].addr < 100) {
+		if (!syms[i].name || syms[i].addr < 100) {
```
Comment (ret2libc) [inline libr/bin/p/bin_mach0.c:216]: why this change? I think it slightly changes the behaviour.

Comment (phakeobj) [reply libr/bin/p/bin_mach0.c:216]: When symbols were allocated on the stack `!syms[i].name[0]` was a way of telling that the symbol is an empty string. Now, that the symbols are allocated on the heap, we don't have empty strings - but NULL pointers. In the case where `syms[i].name == NULL`, wouldn't `syms[i].name[0]` result in a segfault?

Comment (ret2libc) [reply libr/bin/p/bin_mach0.c:216]: yes, but can you have `syms[i] != NULL && syms[i].name[0] == '\0'`?

File: libr/bin/format/mach0/mach0.c (hunk 7 above)
Comment (ret2libc) [reply libr/bin/format/mach0/mach0.c:2906]: why resolved? Did you push new changes?
